### ISSUE TITLE
[REVIEW]: Bedparse: feature extraction from BED files

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1228]
**Submitting author:** @tleonardi (<a href="http://orcid.org/0000-0002-4449-1863">Tommaso Leonardi</a>)
**Repository:** <a href="https://github.com/tleonardi/bedparse" target ="_blank">https://github.com/tleonardi/bedparse</a>
**Version:** v0.2.2
**Editor:** @brainstorm
**Reviewer:** @Benjamin-Lee
**Archive:** <a href="https://doi.org/10.5281/zenodo.2578820" target="_blank">10.5281/zenodo.2578820</a>

## Status

[![status](http://joss.theoj.org/papers/22763a3b37fde13e548e884edd3221fa/status.svg)](http://joss.theoj.org/papers/22763a3b37fde13e548e884edd3221fa)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/22763a3b37fde13e548e884edd3221fa"><img src="http://joss.theoj.org/papers/22763a3b37fde13e548e884edd3221fa/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/22763a3b37fde13e548e884edd3221fa/status.svg)](http://joss.theoj.org/papers/22763a3b37fde13e548e884edd3221fa)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/tleonardi/bedparse" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@Benjamin-Lee, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @brainstorm know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/tleonardi/bedparse">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.2.0)?
- [x] **Authorship:** Has the submitting author (@tleonardi) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @Benjamin-Lee it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1228 with the following error: 

 Can't find any papers to compile :-(

**Comment by tleonardi** [comment]
@whedon generate pdf from branch paper
--

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1228 with the following error: 

 Can't find any papers to compile :-(

**Comment by tleonardi** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1228 with the following error: 

 Can't find any papers to compile :-(

**Comment by arfon** [comment]
@tleonardi - I'm not sure why Whedon can't pick up the branch properly. I've compiled the paper locally: [10.21105.joss.01228.pdf](https://github.com/openjournals/joss-reviews/files/2825803/10.21105.joss.01228.pdf)

**Comment by brainstorm** [comment]
@whedon start review

**Comment by whedon** [comment]
Can't start a review when the review has already started

**Comment by brainstorm** [comment]
Sure, no worries, thanks @kyleniemeyer !

**Comment by tleonardi** [comment]
> @tleonardi - I'm not sure why Whedon can't pick up the branch properly. I've compiled the paper locally: [10.21105.joss.01228.pdf](https://github.com/openjournals/joss-reviews/files/2825803/10.21105.joss.01228.pdf)

Thanks @arfon!

**Comment by Benjamin-Lee** [comment]
As a quick semi-related note, from reading the paper, you might be interested in citing (or signing!) [The Small Tools Manifesto For Bioinformatics](https://github.com/pjotrp/bioinformatics).

**Comment by brainstorm** [comment]
Thanks @Benjamin-Lee! Do you have any comments for @tleonardi regarding the unchecked review boxes? (statement of need, etc?).

**Comment by tleonardi** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1228 with the following error: 

 Can't find any papers to compile :-(

**Comment by arfon** [comment]
> @whedon generate pdf from branch paper

This feature is very buggy I'm afraid and fails for some repositories (looks like yours is one of them).

**Comment by arfon** [comment]
Here's another version I compiled locally: [10.21105.joss.01228.pdf](https://github.com/openjournals/joss-reviews/files/2851159/10.21105.joss.01228.pdf)

**Comment by tleonardi** [comment]
Hi @brainstorm, I've pushed some edits in response to the comments that @Benjamin-Lee made in the bedparse [repo](https://github.com/tleonardi/bedparse/issues). I've also added the citation to the Manifesto and the community guidelines.
@Benjamin-Lee, do you think I should add a direct statement of need? Any comments on how to best address this point?

@arfon, thanks for compiling locally! Let me know if/how I can help to debug the issue!

**Comment by tleonardi** [comment]
@arfon I hadn't pushed the last changes, do you mind compiling the pdf again when you have time? Sorry for the trouble..

**Comment by arfon** [comment]
Here you go: [10.21105.joss.01228.pdf](https://github.com/openjournals/joss-reviews/files/2851544/10.21105.joss.01228.pdf)

**Comment by arfon** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01228/joss.01228/10.21105.joss.01228.pdf)

**Comment by arfon** [comment]
@tleonardi - I fixed the bug :-)

**Comment by tleonardi** [comment]
Brilliant, thanks!

**Comment by Benjamin-Lee** [comment]
Please note that the current unchecked boxes (functionality and performance) are just due to the fact that I haven't had a chance to review those aspects, not due to any issues I've identified with the software.

**Comment by tleonardi** [comment]
Hi @brainstorm, I have added a small feature to bedparse and done a couple of changes following @Benjamin-Lee's feedback. I'd like to make a new release because I have other project's depending on bedparse that require this new feature. Since there's a "version" checkbox, I was wondering whether it's ok for me to make a release at this stage...

**Comment by brainstorm** [comment]
Yes it should be ok as long as @Benjamin-Lee reviews the changes accordingly.

**Comment by tleonardi** [comment]
Great, thanks. I've released v0.2.1.
Actually, I've just now noticed that @Benjamin-Lee ticked all the boxes above. If he's happy with the changes this could become the final release once he is done with the review?

**Comment by Benjamin-Lee** [comment]
I just looked over the changes and they look good.

I also just noticed that you say `Bedparse is a simple python module and CLI tool` in the paper and say it can be imported but there aren't any docs about that. Since it can be imported, there need to be API-level docs. 

In addition, I have a few copy edits, which I submitted as a [PR](https://github.com/tleonardi/bedparse/pull/16).

**Comment by tleonardi** [comment]
Thanks @Benjamin-Lee for pointing out the missing docs. I've written the documentation as docstrings and compiled it with sphinx autodoc. It has it's own [page](https://bedparse.readthedocs.io/en/devel/bedparse.bedline.html) on Read the Docs and simple [example](https://bedparse.readthedocs.io/en/devel/Tutorial.html#apis) in the tutorial.
I've also merged the paper's PR.
Since I've edited the code to complete the docstrings I'll make a new release when the review process is completed.

**Comment by tleonardi** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01228/joss.01228/10.21105.joss.01228.pdf)

**Comment by Benjamin-Lee** [comment]
So sorry, I was looking at the master branch and didn't notice that the devel one has the docs. Is it possible to get method-level examples in the API docs using `:Example:`? (As an aside, you might be interested in [`napoleon`](http://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html). which makes writing sphinx docstrings more pleasant)

Also, what is the reasoning behind using `pprint` and `print` methods over the `__str__()` and `__repr__()` methods and the built-in print function? There seems to be some code duplication between `__str__()` and `__print()` but it may be that I am missing something obvious (like being on the right branch).

The code is well commented and tested. You might be interested in using an automatic formatter such as [black](https://github.com/ambv/black) or [autopep8](https://github.com/hhatto/autopep8) to keep the style consistent and a tool such as [CodeFactor](https://www.codefactor.io) or [Codacy](https://www.codacy.com) for automated code review. 

One minor issue is that `translateChr` shadows the built-in `all` keyword, which [can lead to bad things happening](https://docs.quantifiedcode.com/python-anti-patterns/correctness/assigning_to_builtin.html).

These are small quibbles, however. This tool is useful, documented, tested, and tested. @tleonardi, thank you for your contribution to the bioinformatics community!

@brainstorm and @arfon, the paper is ready to go. In the proof, the URL for the small tools manifesto overruns the edge of the page (see below). Also, the use of the month in one of the citations but not the others is a little bit jarring. These seem to be template-related issues. 
<img width="797" alt="image" src="https://user-images.githubusercontent.com/13618205/53370817-b92b4900-391c-11e9-989f-d2a0663e38f1.png">

**Comment by arfon** [comment]
@Benjamin-Lee @tleonardi this PR should fix the reference formatting: https://github.com/tleonardi/bedparse/pull/17

**Comment by tleonardi** [comment]
No need to be sorry @Benjamin-Lee, I only wrote the docs after you pointed out they were missing! :)
Thanks for the careful review and the new suggestions. I have just renamed the `all` argument in `translateChr()`, thanks for spotting it!
Regarding the  `pprint` and `print` methods, you are right, I should get rid of them. A while back they made sense, but I have been meaning to replace them with  `__str__()` and `__repr__()` for a while... the only complication is that I have quite a few pieces of code in other projects that use those functions, so I have just been keeping them around for backward compatibility. I have opened an [issue](https://github.com/tleonardi/bedparse/issues/18#issue-414539816), I'll refactor as soon as I get a chance.
As for the examples in the docstrings, I'm working on it now.

Thanks @arfon for the PR! It's now merged.

**Comment by tleonardi** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01228/joss.01228/10.21105.joss.01228.pdf)

**Comment by tleonardi** [comment]
Ok, I have added the examples for the API methods in the docstrings. I've also switched to [`napoleon`](http://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html). Thanks @Benjamin-Lee for the tip, the syntax is much more readble!

**Comment by tleonardi** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01228/joss.01228/10.21105.joss.01228.pdf)

**Comment by tleonardi** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01228/joss.01228/10.21105.joss.01228.pdf)

**Comment by tleonardi** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01228/joss.01228/10.21105.joss.01228.pdf)

**Comment by tleonardi** [comment]
Ok, this last one is the version incorporating all the changes suggested by @Benjamin-Lee.
@brainstorm, is there anything else required on my side? I'm ready to make a new release with the last batch of changes.

**Comment by brainstorm** [comment]
Looks good to me, feel free to do that release, pack it and return with a zenodo doi for archival.

**Comment by tleonardi** [comment]
Ok, I've just release v0.2.2.
DOI: [10.5281/zenodo.2578820](https://doi.org/10.5281/zenodo.2578820)
thanks!

**Comment by brainstorm** [comment]
@whedon set 10.5281/zenodo.2578820 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.2578820" target="_blank">10.5281/zenodo.2578820</a> is the archive.

**Comment by brainstorm** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1228 with the following error: 

 Can't find any papers to compile :-(

**Comment by arfon** [comment]
> PDF failed to compile for issue #1228 with the following error:

@tleonardi @brainstorm - the paper must be in the `master` branch for us to accept the submission.

**Comment by tleonardi** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01228/joss.01228/10.21105.joss.01228.pdf)

**Comment by tleonardi** [comment]
@arfon thanks for pointing it out, I had missed this!

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- http://doi.org/10.1101/gr.229102 is OK
- http://doi.org/10.1093/nar/gkx1098 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by tleonardi** [comment]
@arfon It looks like whedon is not picking up the second reference (Prins 2014). Is this normal?

**Comment by arfon** [comment]
> @arfon It looks like whedon is not picking up the second reference (Prins 2014). Is this normal?

Yeah, I think we only check article DOIs at this point.

**Comment by tleonardi** [comment]
Oh, ok, thanks for the clarification ;)

**Comment by brainstorm** [comment]
After another read/checking, it should be ready to go, @arfon, all done here, thanks everyone 👍

**Comment by danielskatz** [comment]
I’ll take care of this in a couple of hours.

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/531

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/531, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- http://doi.org/10.1101/gr.229102 is OK
- http://doi.org/10.1093/nar/gkx1098 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/532
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01228](https://doi.org/10.21105/joss.01228)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by danielskatz** [comment]
Thanks for editing @brainstorm, and for reviewing, @Benjamin-Lee!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.01228/status.svg)](https://doi.org/10.21105/joss.01228)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01228">
  <img src="http://joss.theoj.org/papers/10.21105/joss.01228/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.01228/status.svg
   :target: https://doi.org/10.21105/joss.01228
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.01228/status.svg)](https://doi.org/10.21105/joss.01228)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=

**Comment by tleonardi** [comment]
Great, thank you everyone! In particular thanks to @Benjamin-Lee for reviewing and helping to improve the paper and to @brainstorm for editing!
